### DISCUSSION TITLE
Use Leds-object with hex-value and mqtt

### DISCUSSION THREAD

**Discussion by Lumpi-125** [arendst/Tasmota#20863] [Q&A]
Hi, 
i got a sonoff tx ultimate with ws2812 LEDs. Now i want to play custom LED sequences which are triggert by a json payload over mqtt. Im trying to use the leds-object with set_pixel_color. But i can not use the the color value, because it is a string ("0x00FF00").
Can you please help me? 

This is how i got so far:

`#-
LED WS2812  ( https://tasmota.github.io/docs/Berry_Addressable-LED/ )

     20 21 22 23 24 25 26
19                                    27
18                                    00
17                                    01
16                                    02
15                                    03
14                                    04
13                                    05
     12 11 10 09 08 07 06

-#

#mqtt JSON payload von iobroker -> {"sequence" : 1, "color" : "0x000000"} 

def my_show(topic, idx, payload, bindata)
    
    print("my_show ist gestartet")
    print(payload)

    import json

    var sequenceMemory =	[[[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
                                                  [0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0],
                                                  [0,1,1,1,0,0,0,0,1,1,1,0,0,0,0,1,1,1,0,0,0,0,1,1,1,0,0,0],
                                                  [1,1,1,1,1,0,0,1,1,1,1,1,0,0,1,1,1,1,1,0,0,1,1,1,1,1,0,0],
                                                  [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]],
                                                 [[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
                                                  [0,0,0,0,0,1,1,0,0,0,0,0,1,1,0,0,0,0,0,1,1,0,0,0,0,0,1,1],
                                                  [1,0,0,0,1,1,1,1,0,0,0,1,1,1,1,0,0,0,1,1,1,1,0,0,0,1,1,1],
                                                  [1,1,0,1,1,1,1,1,1,0,1,1,1,1,1,1,0,1,1,1,1,1,1,0,1,1,1,1],
                                                  [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]]]
			  	  

    var current_picture = [0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000]	#aktuelle Anzeige
    var next_picture = [0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000,0x000000]	#nächste Anzeige        
    var my_pixel = Leds(28, 13, Leds.WS2812_GRB)

    var dataJSON = json.load(payload)
    var color = dataJSON["color"]

    var jj = my_pixel.pixel_count() - 1  
    while jj >= 0	#schreibe aktuelle LED Status in current_picture
	current_picture[jj] = my_pixel.get_pixel_color(jj)
	jj = jj - 1
    end

    var ii = size(sequenceMemory[dataJSON["sequence"]])  #ermittle die Frames aus sequenceMemory

    var kk = 0
    while kk < ii
        jj = my_pixel.pixel_count() - 1
        while jj >= 0

	    if sequenceMemory[dataJSON["sequence"]][kk][jj] == 0     #definiere den nächsten zustand der LED
		next_picture[jj] = current_picture[jj]
	    elif sequenceMemory[dataJSON["sequence"]][kk][jj] == 1
	    	next_picture[jj] = color     #when i change color to a maual hex value, then it works
	    else
		next_picture[jj] = 0x000000
	    end

	    my_pixel.set_pixel_color(jj, next_picture[jj])
	    
	    jj = jj - 1
        end
	current_picture = next_picture
	my_pixel.show()
	tasmota.delay(100)
	kk = kk + 1 
    end
end

import mqtt 
mqtt.subscribe("custom/txultimate_bedroom/SEQUENCE", my_show)`

The console says this:
![image](https://github.com/arendst/Tasmota/assets/125072450/bd8007ea-152a-477f-8e31-e9c4011a5cdd)

**Comment by sfromis** [comment]
> But i can not use the the color value, because it is a string ("0x00FF00")

You can convert the string value to an integer `int(color)`

**Reply by Lumpi-125** [reply]
oh man, its that easy. It works, thank you!

**Comment by s-hadinger** [comment]
Be careful when using arrays of arrays of arrays. This is very inefficient when it comes to memory usage.

For example:
```berry
var sequenceMemory =	[[[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
                                              [0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0],
                                              [0,1,1,1,0,0,0,0,1,1,1,0,0,0,0,1,1,1,0,0,0,0,1,1,1,0,0,0],
                                              [1,1,1,1,1,0,0,1,1,1,1,1,0,0,1,1,1,1,1,0,0,1,1,1,1,1,0,0],
                                              [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]],
                                             [[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
                                              [0,0,0,0,0,1,1,0,0,0,0,0,1,1,0,0,0,0,0,1,1,0,0,0,0,0,1,1],
                                              [1,0,0,0,1,1,1,1,0,0,0,1,1,1,1,0,0,0,1,1,1,1,0,0,0,1,1,1],
                                              [1,1,0,1,1,1,1,1,1,0,1,1,1,1,1,1,0,1,1,1,1,1,1,0,1,1,1,1],
                                              [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]]]
```

consumes 2975 bytes of RAM.

You can try more compact encoding using `bytes()`

**Reply by sfromis** [reply]
And with loops being a factor with meaning for speed, you could tune things by putting `sequenceMemory[dataJSON["sequence"]]` into a variable once in the beginning, instead of repeatedly.

A much more compact form could also be to have each 28-bit value as a number, where you could pick off bits by shifting per iteration.

**Reply by Lumpi-125** [reply]
Thanks for checking my code again! I will try to improve it tonight.

**Reply by sfromis** [reply]
Also beware that using `tasmota.delay(100)` is a bad practice, your functions should return as quickly as possible, and within 50 ms. Tasmota is unpresponsive while the code is running, including during a `delay`. To manage timed loops, you should instead be using `tasmota.set_timer` to schedule a next iteration.

More generally, your use case looks like a candidate for the animation framework, which also has functions for managing flow through update cycles.

**Reply by Lumpi-125** [reply]
I read about the animation framework but thought it didn't fit. When I change from one light scene in the room to another, I want the sequence to change from the current color to the next. The sequences depend on the mqtt payload and the current status of the LEDs. With the animation framework I have to define the frames before in terms of color and brightness, right?

**Reply by sfromis** [reply]
I can't see much about your use case outside of the example with predefined frames. As animations commonly have dynamic variations in color and brightness per pixels, I'm not sure what you mean by such being predefined. Anyway, I'm not trying to suggest that it would be a direct fit. My main point was to warn against `delay`.

**Reply by Lumpi-125** [reply]
I'm pretty new to the Tasmota Berry game and I'm very grateful for your tips! If I could use the animation framework, I would really appreciate it. I think with my specific code I will reach my limits with complicated sequences.
I made a small example of what a sequence should look like. In the example, the color should change from blue ("old" color) to green and remain green until a new mqtt command is sent.
Do you think this can be solved with the animation framework?

![image](https://github.com/arendst/Tasmota/assets/125072450/44972556-a045-4401-a568-72c3f6acbad7)

**Reply by Lumpi-125** [reply]
Now i tryed to change my hole system for playing an animation. I calculate the animation in the iobroker and send it to tasmota.

#-
LED WS2812  ( https://tasmota.github.io/docs/Berry_Addressable-LED/ )

  14 15 16 17 18 19 1A
13                    1B
12                    00
11                    01
10                    02
0F                    03
0E                    04
0D                    05
  0C 0B 0A 09 08 07 06


                                                "frame 1"            "frame2"
payload {"frame_time": milsec, "sequence":[[FF00FF00,FF00FF00],[FF00FF00,FF00FF00]]}

  1.Byte is number of pixel PP
2-4.Byte is pixel color RRGGBB

          --> "PPRRGGBB" <--

if the frame contains no new pixel information, the old color will be set

-#

def my_show(topic, idx, payload, bindata)
    
    print("my_show started")
   
    var my_pixel = Leds(28, 13, Leds.WS2812_GRB) #define LEDs

    import json
    var dataJSON = json.load(payload)            #convert payload to JSON
    var frame_time = dataJSON["frame_time"]      #parse frame time
    var sequence = dataJSON["sequence"]          #parse sequence

    var frame = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]

    print(frame)

    var ii = size(sequence)-1                    #get numbers of frames
    
    var kk = 0

    def loop(frame_time, kk, ii, frame, sequence)

        print(kk)

        var frame_change = sequence[kk]

        print(frame_change)

        for i: 0 .. (size(frame_change)-1)
            b=bytes(frame_change[i])
            frame[b.get(0,-1)] = b.get(1,-3)
        end
        
        print(frame)

        for i: 0..27
            my_pixel.set_pixel_color(i, frame[i])
        end
        my_pixel.show()
        
        kk = kk + 1

        if kk <= ii                              #check if all frames has been shown
            tasmota.set_timer(frame_time, def() loop(frame_time, kk, ii, frame, sequence) end)   #set timer to start loop again
        else
            return;
        end

    end
    
    loop(frame_time, kk, ii, frame, sequence)    start loop the first time
 
end

import mqtt 
mqtt.subscribe("custom/txultimate_bedroom/SEQUENCE", my_show)

**Comment by Noschvie** [comment]
BTW is there a WLED to Tasmota converter :-) for this example ?

`
{"on":true, "bri":255,"transition":7,"ps":-1,"pl":-1,"ccnf":{"min":1,"max":5,"time":12},"nl":{"on":false,"dur":60,"fade":true,"mode":1,"tbri":0,"rem":-1},"udpn":{"send":false,"recv":true},"lor":0,"mainseg":0,"seg":[{"id":0,"start":0,"stop":50,"len":50,"grp":1,"spc":0,"on":true,"bri":44,"col":[[255,0,0,0],[0,0,0,0],[0,0,0,0]],"fx":60,"sx":155,"ix":246,"pal":0,"sel":true,"rev":false,"mi":false}]}
`

**Reply by Lumpi-125** [reply]
What do you mean with WLED to Tasmota converter? Do you have some documentation?

**Reply by Noschvie** [reply]
[WLED json api](https://kno.wled.ge/interfaces/json-api/)

using an ESP with WLED to control WS2812B LEDs. If there is a simple solution doing that with Tasmota would be fine.
Thanks!

**Reply by sfromis** [reply]
Since we're already off topic, one possible integration between WLED and Tasmota could be to use DDP, as you can configure Tasmota to serve a DDP data stream for neopixel animations, using `Scheme 16`. Default-available with ESP32, and an optional build time feature with ESP8266.

**Reply by Lumpi-125** [reply]
ah i see, but that doesnt work for me. The sonoff TX Ultimate has Touch control, audio, ws2812 and vibration on board. WLED can not control it all on the ESP32. Thanks to blakadder the is a working template for the tasmota!
https://templates.blakadder.com/sonoff_T5-1C-86.html

**Reply by sfromis** [reply]
Maybe my wording was a bit ambiguous, but it was not about having WLED on the device itself. On a device running Tasmota, you can enable the neopixel part to be controlled via a DDP network stream, and WLED has a feature for having DDP strips, instead of local-only.

I was not trying to suggest this for your use case, it was more of a tangent when Noschvie went off topic talking about WLED and Tasmota getting friendly.
